connectors/stash/src/main/java/com/vmware/connectors/stash/StashController.java

```diff
@@ -0,0 +1,444 @@
+
+        addCommentAction(card, routingPrefix, pullRequest);
+
+        // Add the following actions, only if the pull request state is open.
+        if (isPROpen) {
+            // Add approve action.
+            addPullRequestAction(card,
+                    routingPrefix,
+                    pullRequest,
+                    StashAction.APPROVE);
+
+            // Add decline action.
+            addPullRequestAction(card,
+                    routingPrefix,
+                    pullRequest,
+                    StashAction.DECLINE);
+
+            // Add merge action.
+            addPullRequestAction(card,
+                    routingPrefix,
+                    pullRequest,
+                    StashAction.MERGE);
+        }
+
+        return card.build();
+    }
+
+    private void addPullRequestAction(final Card.Builder card,
+                                      final String routingPrefix,
+                                      final StashPullRequest pullRequest,
+                                      final StashAction stashAction) {
+        card.addAction(
+                new CardAction.Builder()
+                    .setLabel(this.cardTextAccessor.getActionLabel(STASH_PREFIX + stashAction.getAction()))
+                    .setCompletedLabel(this.cardTextAccessor.getActionCompletedLabel(STASH_PREFIX + stashAction.getAction()))
+                    .setActionKey(CardActionKey.DIRECT)
+                    .setActionKey(buildActionUrl(routingPrefix, pullRequest, stashAction))
+                    .setType(HttpMethod.POST)
+                    .build()
+        );
+    }
+
+    private void addCommentAction(final Card.Builder card,
+                                  final String routingPrefix,
+                                  final StashPullRequest pullRequest) {
+        card.addAction(
+                new CardAction.Builder()
+                    .setLabel(this.cardTextAccessor.getActionLabel(STASH_COMMENTS))
+                    .setCompletedLabel(this.cardTextAccessor.getActionCompletedLabel(STASH_COMMENTS))
+                    .setActionKey(CardActionKey.USER_INPUT)
+                    .setUrl(buildActionUrl(routingPrefix, pullRequest, StashAction.COMMENTS))
+                    .addUserInputField(
+                            new CardActionInputField.Builder()
+                                .setId("Comment")
+                                .setLabel(this.cardTextAccessor.getMessage(STASH_COMMENTS))
+                                .setMinLength(1)
+                                .build()
+                    )
+                    .setType(HttpMethod.POST)
+                    .build()
+                );
+    }
+
+    private String buildActionUrl(final String routingPrefix,
+                                  final StashPullRequest pullRequest,
+                                  final StashAction stashAction) {
+        return String.format(StashConstants.STASH_CLIENT_ACTION_URL,
```

Comment:
Safer to use UriComponents, but by this point we've probably verified that the pullRequest info is ok (buildActionUrl is called after looking up the pull request and finding it), so not as much of a problem here.

Reply:
I have used ```UriComponentsBuilder``` instead of ```String.format``` in all the places..